### DISCUSSION TITLE
ObservableProperty: Additional source generated method to allow binding with IObservables (inspired by Flux/Redux)

### DISCUSSION THREAD

**Discussion by aseminjakiw** [CommunityToolkit/dotnet#792] [Q&A]
### TLDR;

I have this view model:

```csharp
public partial class MyViewModel : ObservableObject
{
    [ObservableProperty] private int? _selectedIndex;
}
```

And I would like to have a source generator to generate a method like this, additionaly to the current source generators:
```csharp
// auto generated
public partial class MyViewModel
{
    partial void SetSelectedIndex(int? value)
    {
        if (!global::System.Collections.Generic.EqualityComparer<int?>.Default.Equals(_selectedIndex, value))
        {
            OnPropertyChanging(global::CommunityToolkit.Mvvm.ComponentModel.__Internals.__KnownINotifyPropertyChangingArgs.SelectedIndex);
            _selectedIndex= value;
            OnPropertyChanged(global::CommunityToolkit.Mvvm.ComponentModel.__Internals.__KnownINotifyPropertyChangedArgs.SelectedIndex);
        }
    }
}
```
This is essentially the same as the generated property setter but without the `On***Changeing` and `On***Changed` calls.

### Background
For some time I want to use a Flux/Redux like approach in an XAML environment and MVVM'ish approach. For this I need to consume an `IObsevable<T>`. The challenge is that you need a way to tell the UI that a value changed but at the same time allow for two-way binding scenarios (e.g. SelectedItem in a collection). The generated `On***Changed` methods are great to collect the changes coming from the UI and initiate some action. But if we use the same property setter for changes coming from the internal application state you are at risk to produce an infinity loop. 

So I would like to have a "three-way property" with 
- getter
- setter coming from UI, invoking `On***Changed` and (optionally) raising `INotifyPropertyChanged`
- setter coming from code, raising `INotifyPropertyChanged` to inform UI about changes

With this in place you could write your view model code like this:

```csharp
public record State(int? SelectedIndex, ...);

public partial class MyViewModel : DisposableViewModelBase // DisposableViewModelBase inherits from ObservableObject
{
    [ObservableProperty] private int? _selectedIndex;

    public MyViewModel(IStore<State> store) // IStore implements IObservable<State>
    {
        store.Select(x => x.SelectedIndex).BindProp(this, SetSelectedIndex); //`this` is required to handle releasing the subscription
    }

    partial void OnSetSelectedIndexChanged(int? value) // two-way binding is still working and does not trigger an infinity loop
    {
        _store.Dispatch(new SelectItemAction(value));
    }
}
```

### Questions

1. Is there a change that this will make it into community toolkit (probably not)
2. Can you give me some advice to build this on my own?
3. Do you have a better idea or approach?
